docs/designs/anonlink-upload-data-plan.md

```diff
@@ -0,0 +1,424 @@
+will authenticate with the object store, and ingest the data directly into its database.
+
+In the initial version only binary encoding data can be provided in an external file. This is expected to change after
+the addition of a binary blocking format. 
+
+**Alternative** the backend could copy data to our existing configured object store and then ingest. This method
+doubles up on the data storage and thus wasn't used.
+
+The backend may fail when attempting to pull the data, the credentials may be invalid, the objects may not exist
+at the given path, the external object store could be unavailable or unreachable. In all cases the backend will
+gracefully fail and mark the project as **failed**.
+
+If possible the backend should carry out a quick sanity check during the client's call to the upload endpoint - to 
+verify the object store is up and that the credentials are valid to retrieve an object.
+
+**Future version**: With the addition of a new per data provider `/status` api, the data provider could see any issues
+with their upload and address them without the whole project being marked as a failure.
+
+ 
+### Client Side Specification 
+
+The client side implementation will be in `anonlink-client`, there will be both a public, documented, Python API as well
+as an implementation via the command line tool.
+
+Uploading will be implemented using either [`boto3`](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-uploading-files.html) or 
+[MinIO](https://docs.min.io/docs/python-client-api-reference) at the implementors discretion.
+
+In the default case, uploading via the client tool will involve making the three network requests in sequence: 
+
+- Retrieving temporary object store credentials.
+- Uploading encodings to the object store.
+- Informing the anonlink entity service of the upload.
+
+
+If the user already has object store credentials (e.g. for `S3`) they can upload without having to request
+temporary credentials. The anonlink client tool aims to deal with this transparently.
+
+#### Example client tool usage
+
+The default case using temporary credentials for our object store:
+
+```sh
+$ anonlink upload mydata.csv <AUTH params etc>
+```
+
+Where the user wants our client to upload to their own bucket, optionally providing the AWS credential profile to use:
+
+```sh
+$ anonlink upload mydata.csv [--profile easd] --upload-to=s3://my-own-bucket/mydata.csv <AUTH params etc>
+```
+
+If the user wants to use already uploaded data, they will have to either explicitly provide the anonlink entity service
+with credentials that are appropriate to share with the service, or explicitly request temporary credentials. 
+
+```sh
+$ anonlink upload s3://my-own-bucket/mydata.csv [--profile easd] --request-read-only-credentials \
+           <AUTH params etc>
+```
+
+or 
+```sh
+$ anonlink upload s3://my-own-bucket/mydata.csv [--profile easd] --share-aws-credentials-with-server \
+           <AUTH params etc>
```

Comment:
This are a lot of different ways to upload data, all under the same command. That might lead to a long list of conditionals in the code, also the docs for the command will be rather long to cover all those cases. 
How about we split the cases where we link to already uploaded data and put them into a different command?

Reply:
That's a good idea, I've added `anonlink serverpull` for that second use case. I'm thinking it might be a stretch to implement that in the first revision though.

Reply:
that's fine. I think we should aim to implement the most common use case and it's probably fine to leave the others in the planning stage.